### PR TITLE
Layers for visual attention.

### PR DESCRIPTION
**Pull request by timmoon10**
A `stop_gradient_layer` blocks error signals from passing through during backprop (compare with https://www.tensorflow.org/api_docs/python/tf/stop_gradient). I've also implemented entry-wise min and max layers.

@samadejacobs TensorFlow includes GANs as an example use-case for `stop_gradient`.

### CODE REVIEW

File: include/lbann/layers/transform/stop_gradient.hpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    }
+#endif
+  }
+
+  void fp_compute() override {
+    El::LockedView(get_activations(), get_prev_activations());
+  }
+
+  void bp_compute() override {
+    // El::Zero(get_error_signals());
```
Comment (samadejacobs) [inline include/lbann/layers/transform/stop_gradient.hpp]: This assumes error_signals matrix/container
 is initialized to Zero? Remove commented out code if not needed.

File: include/lbann/layers/transform/max.hpp [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        LBANN_ERROR(err.str());
+      }
+    }
+  }
+
+  void fp_compute() override {
+    const int num_parents = get_num_parents();
+    auto& output = get_activations();
+    switch (num_parents) {
+    case 0: El::Zero(output); break;
```
Comment (samadejacobs) [inline include/lbann/layers/transform/max.hpp]: Can we enforce that min and max layers have at least one parent?

File: include/lbann/layers/transform/stop_gradient.hpp (hunk 1 above)
Comment (timmoon10) [reply include/lbann/layers/transform/stop_gradient.hpp]: This is for convenience once we get around to implementing #347.

File: include/lbann/layers/transform/max.hpp (hunk 2 above)
Comment (timmoon10) [reply include/lbann/layers/transform/max.hpp]: That sounds reasonable. I'll add that restriction.

**Comment by timmoon10** [comment]
Freezing layers completely halts the optimization process for a given set of weights. stop_gradient layers prevent certain error signals from affecting the optimization, but doesn't stop it entirely. If a given layer has only one output, it will compute zero gradients, which will affect the optimizer state. This is more useful when a layer has multiple outputs.
